pkg/params/params.go

```diff
@@ -0,0 +1,225 @@
+	Description string `json:"description"`
+
+	// IsMandatory will be considered when validating; if the param has no value assigned and
+	// also no DefaultValue is set, validation will fail
+	IsMandatory bool `json:"isMandatory"`
+
+	// Tags can be used to skip parameters not needed for a specific environment
+	Tags []string `json:"tags"`
+
+	// Validator is an optional function that will be called upon validation; may or may
+	// not be called in user interfaces. Setting TypeHint is preferred, but can also be used
+	// in combination with the Validator.
+	Validator ParamValidator `json:"-"`
+
+	// TypeHint is the preferred way to set the type of this parameter as it will invoke a
+	// matching validator automatically; if unset, a value of "string" is assumed
+	TypeHint TypeHint `json:"type"`
+
+	// PossibleValues holds all possible values for this parameter and will be considered
+	// when validating
+	PossibleValues []string `json:"possibleValues"`
+
+	value    string
+	assigned bool
+	Value
+}
+
+// GetTitle returns a human friendly title of the field; is no Title has been specified,
+// the Key will be used with the first letter upper-cased
+func (p *Param) GetTitle() string {
+	if p.Title != "" {
+		return p.Title
+	}
+	return cases.Title(language.English).String(p.Key)
+}
+
+// PreValidate validates a string against the given parameter
+func (p *Param) PreValidate(value string) error {
+	if value != "" {
+		if len(p.PossibleValues) > 0 {
+			for _, v := range p.PossibleValues {
+				if v == value {
+					return nil
+				}
+			}
+			return fmt.Errorf("invalid value %q as %q: valid values are: %s", value, p.Key, strings.Join(p.PossibleValues, ", "))
+		}
+		if validator, ok := typeHintValidators[p.TypeHint]; ok {
+			err := validator(value)
+			if err != nil {
+				return fmt.Errorf("invalid value %q as %q: %w", value, p.Key, err)
+			}
+		}
+		if p.Validator != nil {
+			err := p.Validator(value)
+			if err != nil {
+				return fmt.Errorf("invalid value %q as %q: %w", value, p.Key, err)
+			}
+		}
+	} else if p.IsMandatory {
+		return fmt.Errorf("expected value for %q", p.Key)
+	}
+	return nil
+}
+
+// Validate validates an already set value
+func (p *Param) Validate() error {
+	return p.PreValidate(p.String())
+}
+
+func (p *Params) AddParams(other Params) {
+	for _, v := range other {
+		*p = append(*p, v)
+	}
+}
+
+func (p *Params) AddParam(key, value string) {
+	*p = append(*p, &Param{
+		Key:      key,
+		value:    value,
+		assigned: true,
+	})
+}
+
+func (p *Params) ParamMap() (res ParamMap) {
+	res = make(ParamMap)
+	for _, v := range *p {
+		res[v.Key] = v.String()
+	}
+	return
+}
+
+func (p *Params) Validate() error {
+	for _, param := range *p {
+		if err := param.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Get returns the value of the parameter with the given key or an empty string
+func (p *Params) Get(key string) string {
```

Comment:
This is also about avoiding additional error handling. Usually validation of the fields happens beforehand and errors out if a param is not filled correctly, so this wouldn't be called in the first place.

For example, the gadgets can access the stored parameter values upon start, but they wouldn't access it in the first place, because if validation failed, the gadget wouldn't have been started at all.